crates/oxc_parser/src/modifiers.rs

```diff
@@ -102,39 +124,125 @@ pub struct Modifier {
-        Self(Some(modifiers))
+    /// Create a new set of modifiers
+    ///
+    /// # Invariants
+    /// `flags` must correctly reflect the [`ModifierKind`]s within
+    ///  `modifiers`. E.g., if `modifiers` is empty, then so is `flags``.
+    #[must_use]
+    fn new(modifiers: Vec<'a, Modifier>, flags: ModifierFlags) -> Self {
+        if modifiers.is_empty() {
+            debug_assert!(flags.is_empty());
+            Self::empty()
+        } else {
+            // todo: loop over `modifiers` and check integrity of `flags`
+            Self { modifiers: Some(modifiers), flags }
+        }
     }
 
     pub fn empty() -> Self {
-        Self(None)
+        Self { modifiers: None, flags: ModifierFlags::empty() }
+    }
+
+    /// Returns `true` if there are no [`Modifier`]s.
+    pub fn is_empty(&self) -> bool {
+        self.modifiers.is_none()
+    }
+
+    pub fn len(&self) -> usize {
+        self.modifiers.as_ref().map_or(0, |modifiers| modifiers.len())
     }
 
     pub fn contains(&self, target: ModifierKind) -> bool {
-        self.0
-            .as_ref()
-            .map_or(false, |modifiers| modifiers.iter().any(|modifier| modifier.kind == target))
+        self.flags.contains(target.into())
     }
 
     pub fn iter(&self) -> impl Iterator<Item = &Modifier> + '_ {
-        self.0.as_ref().into_iter().flat_map(|modifiers| modifiers.iter())
+        self.modifiers.as_ref().into_iter().flat_map(|modifiers| modifiers.iter())
+    }
+
+    pub fn accessibility(&self) -> Option<TSAccessibility> {
+        self.flags.accessibility()
+    }
+
+    #[inline]
+    pub fn contains_async(&self) -> bool {
+        self.flags.contains(ModifierFlags::ASYNC)
     }
 
-    pub fn is_contains_const(&self) -> bool {
-        self.contains(ModifierKind::Const)
+    #[inline]
+    pub fn contains_const(&self) -> bool {
+        self.flags.contains(ModifierFlags::CONST)
     }
 
-    pub fn is_contains_declare(&self) -> bool {
-        self.contains(ModifierKind::Declare)
+    #[inline]
+    pub fn contains_declare(&self) -> bool {
+        self.flags.contains(ModifierFlags::DECLARE)
     }
 
-    pub fn is_contains_abstract(&self) -> bool {
-        self.contains(ModifierKind::Abstract)
+    #[inline]
+    pub fn contains_abstract(&self) -> bool {
+        self.flags.contains(ModifierFlags::ABSTRACT)
+    }
+
+    #[inline]
+    pub fn contains_readonly(&self) -> bool {
+        self.flags.contains(ModifierFlags::READONLY)
+    }
+
+    #[inline]
+    pub fn contains_override(&self) -> bool {
+        self.flags.contains(ModifierFlags::OVERRIDE)
+    }
+}
+
+impl GetSpan for Modifiers<'_> {
+    fn span(&self) -> Span {
+        let Some(modifiers) = &self.modifiers else { return SPAN };
+        // SAFETY: One of Modifier's invariants is that Some(modifiers) always
+        // contains a non-empty Vec; otherwise it must be `None`.
+        unsafe {
+            use assert_unchecked::assert_unchecked;
```

Comment:
@Boshen is it OK to use `unsafe` here? Or should I use safe code?

Reply:
`debug_assert` should be fine here?